ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/InputsAndEffects.scala

```diff
@@ -78,6 +80,53 @@ private[kvutils] object InputsAndEffects {
       parties.toList.sorted.map(partyStateKey)
     }
 
+    def addOrUpdateResolvedContractId(
+        key: DamlContractKey,
+        contractId: Option[Value.ContractId]): Unit =
+      resolvedContractIdsMap.get(key) match {
+        case None => resolvedContractIdsMap += (key -> contractId)
+        case _ => ()
+      }
+
+    def updateMappingWithExercisesNode(exe: Node.NodeExercises[NodeId, Value.ContractId]): Unit = {
+      if (exe.byKey) {
+        exe.key.foreach { keyWithMaintainers =>
```

Comment:
I've added the `if (exe.byKey)` condition to handle scenarios when a choice on a contract is executed by contract id. In such case the contract key is not a part of the committer's input state and the key validation (added in this PR) will fail.

Reply:
Don't we need an `else` clause here? We probably still need to recognize that we are mutating a key even if we exercise using the contract ID.

Unless it's `nonconsuming`, I guess.

Reply:
@SamirTalwar I think we do but how do we add the key to the input state in that case?

Reply:
The key should be present in the exercise node regardless of it being by key or by coid and it is now added in either case, but only if the exercised choice is `consuming`.

Reply:
The contract keys are also added to the submission inputs now, so that the newly added checks in `TransactionCommitter` can find them (and don't erroneously think that the contracts have been archived between participant execution and commit). This also implies a larger read set that has potentially an adverse performance impact in post-execution conflict detection (when using pre-execution): for this reason, as a next step, I'd like to zoom out and consider possible optimizations. In any case, the performance impact should be assessed _before_ merging this PR. CC @gerolf-da